scripts/bert/run_pretraining_hvd.py

```diff
@@ -0,0 +1,235 @@
+"""
+Pre-training Bidirectional Encoder Representations from Transformers
+=========================================================================================
+This example shows how to pre-train a BERT model with Gluon NLP Toolkit.
+@article{devlin2018bert,
+  title={BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding},
+  author={Devlin, Jacob and Chang, Ming-Wei and Lee, Kenton and Toutanova, Kristina},
+  journal={arXiv preprint arXiv:1810.04805},
+  year={2018}
+}
+"""
+
+# coding: utf-8
+
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+# pylint:disable=redefined-outer-name,logging-format-interpolation
+
+import os
+import random
+import logging
+import time
+import numpy as np
+
+import mxnet as mx
+import gluonnlp as nlp
+
+from utils import profile
+from fp16_utils import FP16Trainer
+from pretraining_utils import get_model_loss, get_pretrain_dataset, get_dummy_dataloader
+from pretraining_utils import save_params, split_and_load, log, evaluate, forward, get_argparser
+
+# parser
+parser = get_argparser()
+args = parser.parse_args()
+
+# logging
+level = logging.DEBUG if args.verbose else logging.INFO
+logging.getLogger().setLevel(level)
+logging.info(args)
+os.environ['MXNET_GPU_MEM_POOL_TYPE'] = 'Round'
+
+try:
+    import horovod.mxnet as hvd
+except ImportError:
+    logging.info('horovod must be installed.')
+    exit()
+hvd.init()
+store = None
+num_workers = hvd.size()
+rank = hvd.rank()
+local_rank = hvd.local_rank()
+if not args.use_avg_len and hvd.size() > 1:
+    logging.info('Specifying --use-avg-len and setting --batch_size with the '
+                 'target number of tokens would help improve training throughput.')
+
+def train(data_train, model, nsp_loss, mlm_loss, vocab_size, ctx):
+    """Training function."""
+    hvd.broadcast_parameters(model.collect_params(), root_rank=0)
+
+    mlm_metric = nlp.metric.MaskedAccuracy()
+    nsp_metric = nlp.metric.MaskedAccuracy()
+    mlm_metric.reset()
+    nsp_metric.reset()
+
+    logging.debug('Creating distributed trainer...')
+    lr = args.lr
+    optim_params = {'learning_rate': lr, 'epsilon': 1e-6, 'wd': 0.01}
```

Comment:
Does need to scale the learning rate by number of workers?
```python
lr = lr * num_workers
```

Reply:
There are studies arguing linear scaling rule vs. sqrt scaling rule vs. fixed value for learning rate with multiple machines. See https://arxiv.org/abs/1808.01371 , "Large-Batch Training for LSTM and Beyond", and https://arxiv.org/abs/1706.02677 
The learning rate is left for the user to decide

Reply:
Also FYI https://github.com/horovod/horovod/blob/master/horovod/mxnet/__init__.py#L96